sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/UploadWorker.java

```diff
@@ -0,0 +1,150 @@
+import androidx.concurrent.futures.CallbackToFutureAdapter;
+import androidx.work.Data;
+import androidx.work.ListenableWorker;
+import androidx.work.WorkerParameters;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+/**
+ * Package private.
+ *
+ * {@link ListenableWorker} for performing a single file upload using {@link UploadHandler}.
+ */
+class UploadWorker extends ListenableWorker {
+    private static final String TAG = UploadWorker.class.getSimpleName();
+    // The number of blob blocks to be uploaded in parallel.
+    private int blocksUploadConcurrency;
+    // The key of the blob upload metadata entity describing the file to be uploaded.
+    private long blobUploadId;
+    // A token to signal {@link UploadHandler} that it should be stopped.
+    private UploadStopToken uploadStopToken;
+
+    /**
+     * Create the upload worker.
+     *
+     * {@link TransferClient} won't call this constructor directly instead it enqueue the work along
+     * with criteria to execute it. The {@link androidx.work.WorkManager} call this Ctr when it
+     * identifies that the criteria are satisfied (e.g network is available).
+     *
+     * If criteria configured for the Worker is no longer met while it is executing, then WorkManager
+     * will stop the Worker by calling {@link ListenableWorker#stop()}. Such stopped Worker instance
+     * will be eventually GC-ed. When constraints are again satisfied and if the Worker was stopped
+     * without completion then WorkManager will create and starts new Worker instance.
+     *
+     * @param appContext the context
+     * @param workerParams the input parameters to the worker
+     */
+    public UploadWorker(@NonNull Context appContext,
+                        @NonNull WorkerParameters workerParams) {
+        super(appContext, workerParams);
+        this.blobUploadId
+            = getInputData().getLong(Constants.INPUT_BLOB_UPLOAD_ID_KEY, -1);
+        if (this.blobUploadId <= -1) {
+            throw new IllegalArgumentException("Worker created with no or non-positive input blobUploadId.");
+        }
+        this.blocksUploadConcurrency
+            = getInputData().getInt(Constants.INPUT_BLOCKS_UPLOAD_CONCURRENCY_KEY,
+            Constants.DEFAULT_BLOCKS_UPLOAD_CONCURRENCY);
+        if (this.blocksUploadConcurrency <= 0) {
+            this.blocksUploadConcurrency = Constants.DEFAULT_BLOCKS_UPLOAD_CONCURRENCY;
+        }
+    }
+
+    /**
+     * WorkManager calls startWork() on main-thread, so it is important that the work this method
+     * does is light weight. This method creates an {@link UploadHandler} and delegate the
+     * file upload work.
+     *
+     * @return the future representing the async upload.
+     */
+    @NonNull
+    @Override
+    public ListenableFuture<Result> startWork() {
+        Log.v(TAG, "startWork() called." + this);
+        ListenableFuture<Result> listenableFuture = CallbackToFutureAdapter.getFuture(completer -> {
```

Comment:
Safe to use the lambda syntax here? I think so, just checking.

Reply:
it's safe. Here is the [documentation](https://developer.android.com/reference/androidx/concurrent/futures/CallbackToFutureAdapter) with similar Java usage.